File: Python/8-Graphs(BFS-DFS)/Problems/1-Number-of-Islands/Number-of-Islands.py
"""
Number of Islands Problem:
Given a 2D grid map of '1's (land) and '0's (water), count the number of islands.
An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.

Approaches Implemented:
1. Depth-First Search (DFS) - Recursive
2. Depth-First Search (DFS) - Iterative
3. Breadth-First Search (BFS)
4. Union-Find (Disjoint Set)
5. Flood Fill Algorithm
"""
from collections import deque

# Approach 1: DFS Recursive
def num_islands_dfs_recursive(grid):
    """
    Counts the number of islands in a grid using recursive Depth-First Search (DFS).

    Args:
        grid (List[List[str]]): A 2D list of strings representing the grid ('1' for land, '0' for water).

    Returns:
        int: The number of islands.

    Algorithm:
        1.  If the grid is empty, return 0.
        2.  Initialize `rows` and `cols` to the dimensions of the grid.
        3.  `directions` stores the possible movements (up, down, left, right).
        4.  Define a recursive function `dfs(r, c)`:
            a.  Base case: If the current cell is out of bounds or is water ('0'), return.
            b.  Mark the current cell as visited by setting it to '0'.
            c.  Recursively call `dfs` for the four adjacent cells.
        5.  Initialize `count` to 0.
        6.  Iterate through each cell in the grid:
            a.  If the cell is land ('1'):
                i.  Increment `count`.
                ii. Call `dfs(r, c)` to explore the entire island.
        7.  Return `count`.

    Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
    Space Complexity: O(M * N) in the worst case (when the entire grid is land), due to the recursion depth.
    """
    if not grid:
        return 0

    rows, cols = len(grid), len(grid[0])
    directions = [(0,1), (1,0), (0,-1), (-1,0)]

    def dfs(r, c):
        if r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] == '0':
            return
        grid[r][c] = '0'  # Mark visited
        for dr, dc in directions:
            dfs(r + dr, c + dc)

    count = 0
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                count += 1
                dfs(r, c)
    return count

# Approach 2: DFS Iterative
def num_islands_dfs_iterative(grid):
    """
    Counts the number of islands in a grid using iterative Depth-First Search (DFS) with a stack.

    Args:
        grid (List[List[str]]): A 2D list of strings representing the grid.

    Returns:
        int: The number of islands.

    Algorithm:
        1.  If the grid is empty, return 0.
        2.  Initialize `rows`, `cols`, and `directions`.
        3.  Initialize an empty stack and `count` to 0.
        4.  Iterate through each cell in the grid:
            a.  If the cell is land ('1'):
                i.  Increment `count`.
                ii. Push the cell's coordinates onto the stack.
                iii. While the stack is not empty:
                    1.  Pop a cell's coordinates (x, y) from the stack.
                    2.  If the cell is land ('1'):
                        a.  Mark the cell as visited ('0').
                        b.  For each neighbor:
                            i.  If the neighbor is within bounds and is land ('1'), push its coordinates onto the stack.
        5.  Return `count`.

    Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
    Space Complexity: O(M * N) in the worst case (when the entire grid is land), due to the stack.
    """
    if not grid:
        return 0

    rows, cols = len(grid), len(grid[0])
    directions = [(0,1), (1,0), (0,-1), (-1,0)]
    stack = []
    count = 0

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                count += 1
                stack.append((r, c))
                while stack:
                    x, y = stack.pop()
                    if grid[x][y] == '1':
                        grid[x][y] = '0'  # Mark visited
                        for dr, dc in directions:
                            nr, nc = x + dr, y + dc
                            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '1':
                                stack.append((nr, nc))
    return count

# Approach 3: BFS
def num_islands_bfs(grid):
    """
    Counts the number of islands in a grid using Breadth-First Search (BFS).

    Args:
        grid (List[List[str]]): A 2D list of strings representing the grid.

    Returns:
        int: The number of islands.

    Algorithm:
        1.  If the grid is empty, return 0.
        2.  Initialize `rows`, `cols`, and `directions`.
        3.  Initialize an empty queue and `count` to 0.
        4.  Iterate through each cell in the grid:
            a.  If the cell is land ('1'):
                i.  Increment `count`.
                ii. Enqueue the cell's coordinates.
                iii. While the queue is not empty:
                    1.  Dequeue a cell's coordinates (x, y).
                    2.  If the cell is land ('1'):
                        a.  Mark the cell as visited ('0').
                        b.  For each neighbor:
                            i.  If the neighbor is within bounds and is land ('1'), enqueue its coordinates.
        5.  Return `count`.

    Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
    Space Complexity: O(M * N) in the worst case (when the entire grid is land), due to the queue.
    """
    if not grid:
        return 0

    rows, cols = len(grid), len(grid[0])
    directions = [(0,1), (1,0), (0,-1), (-1,0)]
    queue = deque()
    count = 0

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                count += 1
                queue.append((r, c))
                while queue:
                    x, y = queue.popleft()
                    if grid[x][y] == '1':
                        grid[x][y] = '0'  # Mark visited
                        for dr, dc in directions:
                            nr, nc = x + dr, y + dc
                            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '1':
                                queue.append((nr, nc))
    return count

# Approach 4: Union-Find (Disjoint Set)
class UnionFind:
    """
    Represents a Union-Find (Disjoint Set) data structure.  This class is used to efficiently
    track connected components in the grid.

    Attributes:
        parent (List[int]):  A list where parent[i] stores the parent of element i.
        rank (List[int]): A list where rank[i] stores the rank (height) of the subtree rooted at i.
        count (int): The number of disjoint sets.

    Methods:
        __init__(size): Initializes the Union-Find structure with a given size.
        find(x): Finds the root (representative) of the set that element x belongs to, using path compression.
        union(x, y): Merges the sets containing elements x and y, using union by rank.
    """
    def __init__(self, size):
        """
        Initializes the Union-Find data structure.

        Args:
            size (int): The number of elements.
        """
        self.parent = list(range(size))
        self.rank = [1] * size
        self.count = 0

    def find(self, x):
        """
        Finds the root (representative) of the set that element x belongs to.
        Implements path compression for efficiency.

        Args:
            x (int): The element to find the root of.

        Returns:
            int: The root of the set containing x.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x, y):
        """
        Merges the sets containing elements x and y.
        Implements union by rank for efficiency.

        Args:
            x (int): The first element.
            y (int): The second element.
        """
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1
            self.count -= 1

def num_islands_union_find(grid):
    """
    Counts the number of islands in a grid using the Union-Find (Disjoint Set) algorithm.

    Args:
        grid (List[List[str]]): A 2D list of strings representing the grid.

    Returns:
        int: The number of islands.

    Algorithm:
        1.  If the grid is empty, return 0.
        2.  Initialize `rows` and `cols`.
        3.  Create a UnionFind object with the size of the grid (rows * cols).
        4. `directions` stores the possible movements (right, down).  We only need to check right and down
           because checking left and up would duplicate work.
        5.  Iterate through each cell in the grid:
            a.  If the cell is land ('1'):
                i.  Increment the `count` in the UnionFind object.
                ii. For each neighbor (right and down):
                    1.  If the neighbor is within bounds and is land ('1'):
                        a.  Perform a union operation on the current cell and the neighbor in the UnionFind object.  The cell indices are calculated as r * cols + c.
        6.  Return the `count` from the UnionFind object, which represents the number of disjoint sets (islands).

    Time Complexity: O(M * N * α(M * N)), where α is the inverse Ackermann function, which is very slow-growing and
                     can be considered almost constant for practical input sizes.  Essentially, it's nearly O(M*N).
    Space Complexity: O(M * N) for the UnionFind data structure.
    """
    if not grid:
        return 0

    rows, cols = len(grid), len(grid[0])
    uf = UnionFind(rows * cols)
    directions = [(0,1), (1,0)]

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                uf.count += 1
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '1':
                        uf.union(r * cols + c, nr * cols + nc)
    return uf.count

# Approach 5: Flood Fill Algorithm
def num_islands_flood_fill(grid):
    """
    Counts the number of islands in a grid using the Flood Fill algorithm.  This is very similar
    to the recursive DFS approach, but it's presented here as a distinct approach for completeness.

    Args:
        grid (List[List[str]]): A 2D list of strings representing the grid.

    Returns:
        int: The number of islands.

    Algorithm:
        1.  If the grid is empty, return 0.
        2.  Initialize `rows` and `cols`.
        3.  Define a recursive function `flood_fill(r, c)`:
            a.  Base case: If the current cell is out of bounds or is water ('0'), return.
            b.  Mark the current cell as visited by setting it to '0'.
            c.  Recursively call `flood_fill` for the four adjacent cells.
        4.  Initialize `count` to 0.
        5.  Iterate through each cell in the grid:
            a.  If the cell is land ('1'):
                i.  Increment `count`.
                ii. Call `flood_fill(r, c)` to "flood" the entire island, marking all connected land cells as visited.
        6.  Return `count`.

    Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
    Space Complexity: O(M * N) in the worst case (when the entire grid is land), due to the recursion depth.
    """
    if not grid:
        return 0

    rows, cols = len(grid), len(grid[0])

    def flood_fill(r, c):
        if r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] == '0':
            return
        grid[r][c] = '0'  # Mark visited
        flood_fill(r+1, c)
        flood_fill(r-1, c)
        flood_fill(r, c+1)
        flood_fill(r, c-1)

    count = 0
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                count += 1
                flood_fill(r, c)
    return count

# Test the functions
grid = [
    ['1', '1', '0', '0', '0'],
    ['1', '1', '0', '0', '0'],
    ['0', '0', '1', '0', '0'],
    ['0', '0', '0', '1', '1']
]

print("DFS Recursive:", num_islands_dfs_recursive([row[:] for row in grid]))
print("DFS Iterative:", num_islands_dfs_iterative([row[:] for row in grid]))
print("BFS:", num_islands_bfs([row[:] for row in grid]))
print("Union-Find:", num_islands_union_find([row[:] for row in grid]))
print("Flood Fill:", num_islands_flood_fill([row[:] for row in grid]))

File: C++/8-Graphs(BFS-DFS)/Problems/1-Number-of-Islands/Number-of-Islands.js
/**
 * Given a 2D grid map of '1's (land) and '0's (water), count the number of islands.
 * An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
 * You may assume all four edges of the grid are all surrounded by water.
 *
 * Example 1:
 *
 * Input:
 * [
 * ['1', '1', '1', '1', '0'],
 * ['1', '1', '0', '1', '0'],
 * ['1', '1', '0', '0', '0'],
 * ['0', '0', '0', '0', '0']
 * ]
 * Output: 1
 *
 * Example 2:
 *
 * Input:
 * [
 * ['1', '1', '0', '0', '0'],
 * ['1', '1', '0', '0', '0'],
 * ['0', '0', '1', '0', '0'],
 * ['0', '0', '0', '1', '1']
 * ]
 * Output: 3
 */

/**
 * Approach 1: Depth First Search (DFS)
 *
 * Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
 * Space Complexity: O(M * N) in the worst case (when the grid is filled with '1's), due to the recursion stack.
 */
function numIslandsDFS(grid) {
  if (!grid || grid.length === 0) {
    return 0;
  }

  const rows = grid.length;
  const cols = grid[0].length;
  let numIslands = 0;

  function dfs(row, col) {
    if (row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] === '0') {
      return;
    }

    grid[row][col] = '0'; // Mark the current land as visited
    // Explore adjacent cells
    dfs(row + 1, col);
    dfs(row - 1, col);
    dfs(row, col + 1);
    dfs(row, col - 1);
  }

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === '1') {
        numIslands++;
        dfs(i, j);
      }
    }
  }

  return numIslands;
}

/**
 * Approach 2: Breadth First Search (BFS)
 *
 * Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
 * Space Complexity: O(min(M, N)) in the worst case, due to the queue size.
 */
function numIslandsBFS(grid) {
  if (!grid || grid.length === 0) {
    return 0;
  }

  const rows = grid.length;
  const cols = grid[0].length;
  let numIslands = 0;
  const queue = [];

  const directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === '1') {
        numIslands++;
        queue.push([i, j]);
        grid[i][j] = '0'; // Mark the current land as visited

        while (queue.length > 0) {
          const [currRow, currCol] = queue.shift();

          for (const [deltaRow, deltaCol] of directions) {
            const nextRow = currRow + deltaRow;
            const nextCol = currCol + deltaCol;

            if (
              nextRow >= 0 &&
              nextRow < rows &&
              nextCol >= 0 &&
              nextCol < cols &&
              grid[nextRow][nextCol] === '1'
            ) {
              grid[nextRow][nextCol] = '0'; // Mark the neighbor as visited
              queue.push([nextRow, nextCol]);
            }
          }
        }
      }
    }
  }

  return numIslands;
}

/**
 * Approach 3: Union Find (Disjoint Set)
 *
 * Time Complexity: O(M * N * α(M * N)), where α is the Inverse Ackermann function,
 * which is very slowly growing and can be considered almost constant.
 * Space Complexity: O(M * N)
 */
class UnionFind {
  constructor(grid) {
    this.count = 0;
    this.parent = [];
    this.rank = [];
    const rows = grid.length;
    const cols = grid[0].length;

    for (let i = 0; i < rows; i++) {
      for (let j = 0; j < cols; j++) {
        if (grid[i][j] === '1') {
          this.parent[i * cols + j] = i * cols + j;
          this.count++;
        }
        this.rank[i * cols + j] = 0;
      }
    }
  }

  find(i) {
    if (this.parent[i] !== i) {
      this.parent[i] = this.find(this.parent[i]);
    }
    return this.parent[i];
  }

  union(i, j) {
    const rootP = this.find(i);
    const rootQ = this.find(j);
    if (rootP !== rootQ) {
      if (this.rank[rootP] > this.rank[rootQ]) {
        this.parent[rootQ] = rootP;
      } else if (this.rank[rootP] < this.rank[rootQ]) {
        this.parent[rootP] = rootQ;
      } else {
        this.parent[rootQ] = rootP;
        this.rank[rootP]++;
      }
      this.count--;
    }
  }

  getCount() {
    return this.count;
  }
}

function numIslandsUnionFind(grid) {
  if (!grid || grid.length === 0) {
    return 0;
  }

  const rows = grid.length;
  const cols = grid[0].length;
  const uf = new UnionFind(grid);
  const directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === '1') {
        for (const [deltaRow, deltaCol] of directions) {
          const nextRow = i + deltaRow;
          const nextCol = j + deltaCol;
          if (
            nextRow >= 0 &&
            nextRow < rows &&
            nextCol >= 0 &&
            nextCol < cols &&
            grid[nextRow][nextCol] === '1'
          ) {
            uf.union(i * cols + j, nextRow * cols + nextCol);
          }
        }
      }
    }
  }
  return uf.getCount();
}

/**
 * Approach 4: Iterative Depth First Search (DFS) with Stack
 *
 * Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
 * Space Complexity: O(M * N) in the worst case (when the grid is filled with '1's).
 */
function numIslandsIterativeDFS(grid) {
  if (!grid || grid.length === 0) {
    return 0;
  }

  const rows = grid.length;
  const cols = grid[0].length;
  let numIslands = 0;
  const stack = [];

  const directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === '1') {
        numIslands++;
        stack.push([i, j]);
        grid[i][j] = '0'; // Mark as visited

        while (stack.length > 0) {
          const [currRow, currCol] = stack.pop();

          for (const [deltaRow, deltaCol] of directions) {
            const nextRow = currRow + deltaRow;
            const nextCol = currCol + deltaCol;

            if (
              nextRow >= 0 &&
              nextRow < rows &&
              nextCol >= 0 &&
              nextCol < cols &&
              grid[nextRow][nextCol] === '1'
            ) {
              grid[nextRow][nextCol] = '0'; // Mark as visited
              stack.push([nextRow, nextCol]);
            }
          }
        }
      }
    }
  }
  return numIslands;
}

/**
 * Approach 5: Modified Input (In-place modification of grid) with DFS
 * This approach modifies the input grid directly to mark visited cells.
 *
 * Time Complexity: O(M * N), where M is the number of rows and N is the number of columns.
 * Space Complexity: O(M * N) in the worst case due to the recursion stack.  However, this approach has the advantage of not using additional data structures like a queue or visited set, but it modifies the input.
 */
function numIslandsModifiedInputDFS(grid) {
    if (!grid || grid.length === 0) {
        return 0;
    }

    const rows = grid.length;
    const cols = grid[0].length;
    let numIslands = 0;

    function dfs(row, col) {
        if (row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] === '0') {
            return;
        }

        grid[row][col] = '0'; // Mark the current cell as visited by changing its value
        dfs(row + 1, col);
        dfs(row - 1, col);
        dfs(row, col + 1);
        dfs(row, col - 1);
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (grid[i][j] === '1') {
                numIslands++;
                dfs(i, j);
            }
        }
    }
    return numIslands;
}



// Example Usage:
const grid1 = [
  ['1', '1', '1', '1', '0'],
  ['1', '1', '0', '1', '0'],
  ['1', '1', '0', '0', '0'],
  ['0', '0', '0', '0', '0'],
];

const grid2 = [
  ['1', '1', '0', '0', '0'],
  ['1', '1', '0', '0', '0'],
  ['0', '0', '1', '0', '0'],
  ['0', '0', '0', '1', '1'],
];

const grid3 = [
    ['1', '1', '1'],
    ['0', '1', '0'],
    ['1', '1', '1']
];

const grid4 = [
    ['1', '0', '1', '1', '1'],
    ['1', '0', '1', '0', '1'],
    ['1', '1', '1', '0', '1']
];

const grid5 = [
  ['0', '1', '0'],
  ['1', '0', '1'],
  ['0', '1', '0']
];


console.log("Input Grid 1:", grid1);
console.log("DFS:", numIslandsDFS(grid1)); // Output: 1
console.log("BFS:", numIslandsBFS(grid1)); // Output: 1
console.log("Union Find:", numIslandsUnionFind(grid1)); // Output: 1
console.log("Iterative DFS:", numIslandsIterativeDFS(grid1)); // Output: 1

// IMPORTANT:  Make a copy, because the Modified DFS *changes* the input.
const grid1Copy = grid1.map(row => [...row]);
console.log("Modified Input DFS:", numIslandsModifiedInputDFS(grid1Copy));
console.log("Modified Grid 1:", grid1Copy); // The grid1Copy is now modified.

console.log("\nInput Grid 2:", grid2);
console.log("DFS:", numIslandsDFS(grid2)); // Output: 3
console.log("BFS:", numIslandsBFS(grid2)); // Output: 3
console.log("Union Find:", numIslandsUnionFind(grid2)); // Output: 3
console.log("Iterative DFS:", numIslandsIterativeDFS(grid2)); // Output: 3
const grid2Copy = grid2.map(row => [...row]);
console.log("Modified Input DFS:", numIslandsModifiedInputDFS(grid2Copy));
console.log("Modified Grid 2:", grid2Copy);

console.log("\nInput Grid 3:", grid3);
console.log("DFS:", numIslandsDFS(grid3)); // Output: 1
console.log("BFS:", numIslandsBFS(grid3)); // Output: 1
console.log("Union Find:", numIslandsUnionFind(grid3)); // Output: 1
console.log("Iterative DFS:", numIslandsIterativeDFS(grid3)); // Output: 1
const grid3Copy = grid3.map(row => [...row]);
console.log("Modified Input DFS:", numIslandsModifiedInputDFS(grid3Copy));
console.log("Modified Grid 3:", grid3Copy);

console.log("\nInput Grid 4:", grid4);
console.log("DFS:", numIslandsDFS(grid4)); // Output: 3
console.log("BFS:", numIslandsBFS(grid4)); // Output: 3
console.log("Union Find:", numIslandsUnionFind(grid4)); // Output: 3
console.log("Iterative DFS:", numIslandsIterativeDFS(grid4)); // Output: 3
const grid4Copy = grid4.map(row => [...row]);
console.log("Modified Input DFS:", numIslandsModifiedInputDFS(grid4Copy));
console.log("Modified Grid 4:", grid4Copy);

console.log("\nInput Grid 5:", grid5);
console.log("DFS:", numIslandsDFS(grid5)); // Output = 2
console.log("BFS:", numIslandsBFS(grid5)); // Output = 2
console.log("Union Find:", numIslandsUnionFind(grid5));  // Output: 2
console.log("Iterative DFS:", numIslandsIterativeDFS(grid5)); // Output: 2
const grid5Copy = grid5.map(row => [...row]);
console.log("Modified Input DFS:", numIslandsModifiedInputDFS(grid5Copy));
console.log("Modified Grid 5:", grid5Copy);
